chainer/links/normalization/batch_normalization.py

```diff
@@ -62,32 +67,41 @@ class BatchNormalization(link.Link):
 
     """
 
-    def __init__(self, size, decay=0.9, eps=2e-5, dtype=numpy.float32,
-                 use_gamma=True, use_beta=True,
-                 initial_gamma=None, initial_beta=None):
+    def __init__(self, size=None, decay=0.9, eps=2e-5,
+                 dtype=numpy.float32, use_gamma=True, use_beta=True,
+                 initial_gamma=None, initial_beta=None, axis=1):
         super(BatchNormalization, self).__init__()
-        self.avg_mean = numpy.zeros(size, dtype=dtype)
-        self.register_persistent('avg_mean')
-        self.avg_var = numpy.zeros(size, dtype=dtype)
-        self.register_persistent('avg_var')
+
         self.N = 0
         self.register_persistent('N')
         self.decay = decay
         self.eps = eps
 
+        self._size = size
+        self._axis = (axis,) if isinstance(axis, int) else tuple(axis)
+        self._dtype = dtype
+
         with self.init_scope():
             if use_gamma:
                 if initial_gamma is None:
                     initial_gamma = 1
-                initial_gamma = initializers._get_initializer(initial_gamma)
-                initial_gamma.dtype = dtype
-                self.gamma = variable.Parameter(initial_gamma, size)
+                gamma_initializer = initializers._get_initializer(initial_gamma)
+                gamma_initializer.dtype = self._dtype
+                self.gamma = variable.Parameter(gamma_initializer)
             if use_beta:
                 if initial_beta is None:
                     initial_beta = 0
-                initial_beta = initializers._get_initializer(initial_beta)
-                initial_beta.dtype = dtype
-                self.beta = variable.Parameter(initial_beta, size)
+                beta_initializer = initializers._get_initializer(initial_beta)
+                beta_initializer.dtype = self._dtype
+                self.beta = variable.Parameter(beta_initializer)
+
+    def _initialize_params(self, shape):
+        self.avg_mean = numpy.zeros(shape, dtype=self._dtype)
+        self.register_persistent('avg_mean')
+        self.avg_var = numpy.zeros(shape, dtype=self._dtype)
+        self.register_persistent('avg_var')
+        self.gamma.initialize(shape)
```

Comment:
If `use_gamma` is `False`, the instance does not have `gamma` attribute. So, please check its existence (e.g. with `hasattr`).

Reply:
The current implementation in master branch does not set `gamma` and `beta` if `use_gamma` and `user_beta` is `False`, respectively. I just tried the following code

```python
>> bn = BatchNormalization(16, use_gamma=False, use_beta=False)
>> hasattr(bn, 'beta')
False
>> hasattr(bn, 'gamma')
False
```

Am I missing something here?

Reply:
My bad, I got what you mean. You want me to check if an instance has the attributes in `_initialize_params()`. I will fix what you pointed out as soon as possible. Thanks!

Reply:
Yes, that was what I wanted to say. Sorry for confusing you.